### PR TITLE
qt6: fix broken debug mode and dependency build type overriding

### PR DESCRIPTION
**Pull request by outfoxxed**
## Description of changes

This fixes two major issues with qt6's packaging. They're fairly linked together so this is done as a single PR.

### First issue:
Previously the qt setup hook would force qtbase's cmake and qmake build types on dependent packages, usually causing them to build in Release mode.This is especially bad in development environments but still bad for distribution if you want RelWithDebInfo.

Now the build type can be set on a per package basis, and does not affect dependent packages.

### Second issue:
Previously the debug and developerBuild argument caused *qtbase* to build in `Release` mode without stripping symbols and all modules to build in `Debug` mode, but stripped them, rendering both entirely useless for debugging. Now the debug works at a per-module level and for the whole qt6 set.

In other words `(pkgs.qt6.override { debug = true; }).module` and `pkgs.qt6.module.override { debug = true; }` now work correctly.

It also removes `developerBuild` entirely, as the only difference was adding GDB to path.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable: (there are no tests I could find for qt)
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

I've tested this fully on a local project using qtbase qtdeclarative and qtwayland, and run a full build of qt6.full.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by NickCao** [comment]
I think the qt6 build system already handles the standard `CMAKE_BUILD_TYPE` option, maybe we can just drop all these special handling?

**Comment by outfoxxed** [comment]
> I think the qt6 build system already handles the standard CMAKE_BUILD_TYPE option, maybe we can just drop all these special handling?

Honestly I have no idea.

File: pkgs/development/libraries/qt-6/modules/qtbase.nix [hunk 1]
```diff
@@ -250,7 +245,7 @@ stdenv.mkDerivation rec {
     "-framework GSS"
   ]);
 
-  env.NIX_CFLAGS_COMPILE = "-DNIXPKGS_QT_PLUGIN_PREFIX=\"${qtPluginPrefix}\"";
+  env.NIX_CFLAGS_COMPILE = "-DNIXPKGS_QT_PLUGIN_PREFIX=\"${qtPluginPrefix}\"${if debug then "" else " -DQT_NO_DEBUG"}";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/qt-6/modules/qtbase.nix:248]:
```suggestion
  env.NIX_CFLAGS_COMPILE = "-DNIXPKGS_QT_PLUGIN_PREFIX=\"${qtPluginPrefix}\"${lib.optionalString (!debug) " -DQT_NO_DEBUG"}";
```

File: pkgs/development/libraries/qt-6/qtModule.nix [hunk 2]
```diff
@@ -26,10 +27,17 @@ stdenv.mkDerivation (args // {
     (lib.warnIf (args ? qtInputs) "qt6.qtModule's qtInputs argument is deprecated" args.qtInputs or []) ++
     (args.propagatedBuildInputs or []);
 
+  env.NIX_CFLAGS_COMPILE = let
+    prev = if args ? env && args.env ? NIX_CFLAGS_COMPILE then "${args.env.NIX_CFLAGS_COMPILE} " else "";
+  in "${prev}${if debug then "" else "-DQT_NO_DEBUG"}";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/qt-6/qtModule.nix:32]:
```suggestion
    prev = lib.optionalString (args ? env && args.env ? NIX_CFLAGS_COMPILE) "${args.env.NIX_CFLAGS_COMPILE} ";
  in "${prev}${lib.optonalString (!debug) "-DQT_NO_DEBUG"}";
```

**Comment by SuperSandro2000** [comment]
I cannot really say anything about the actual changes.

**Comment by NickCao** [comment]
Trying out an alternative version in https://github.com/NixOS/nixpkgs/pull/307573. As previously said, the current qt build system should be able to automatically set `-DQT_NO_DEBUG` according to `cmakeBuildType`, just need to double test.

**Comment by NickCao** [comment]
Superseded by https://github.com/NixOS/nixpkgs/pull/307573

**Pull request closed without merging**